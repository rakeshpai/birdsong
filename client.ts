import type { MyServiceType } from './server';

const createService = <T>() => new Proxy({}, {
  get(target, prop) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (...args: any[]) => {
      console.log('Invoking', prop, args);
      return Promise.resolve(3);
    };
  }
}) as T;

const service = createService<MyServiceType>();

const r1 = service.add(1, 2);
const r2 = service.addDates(new Date(), new Date());
const r3 = service.concat('a', 'b');

Promise.all([r1, r2, r3]).then(x => console.log(x));
