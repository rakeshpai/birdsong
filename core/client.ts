export const createClient = <T>() => new Proxy({}, {
  get(target, prop) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (...args: any[]) => {
      // eslint-disable-next-line no-console
      console.log('Invoking', prop, args);
      return Promise.resolve(3);
    };
  }
}) as T;
