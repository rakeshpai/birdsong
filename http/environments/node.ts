import type { IncomingMessage, ServerResponse } from 'http';
import cookie from 'cookie';
import { URL, URLSearchParams } from 'url';
import type { EnvironmentHelpers } from '../types';
import { getMethodDetails } from './helpers';

const nodejs = (request: IncomingMessage, response: ServerResponse): EnvironmentHelpers => {
  const cookies = cookie.parse(request.headers.cookie || '');
  const { searchParams } = new URL(request.url || '');

  const postBody = (() => {
    let body = '';
    request.on('data', chunk => { body += chunk; });
    const result = new Promise(resolve => { request.on('end', () => { resolve(body); }); });
    return async () => {
      const body = await (result as Promise<string>);
      const parsed = new URLSearchParams(body);
      return { method: parsed.get('method'), input: parsed.get('input') };
    };
  })();

  const setCookie = (name: string, value: string) => {
    response.setHeader('Set-Cookie', cookie.serialize(name, value));
  };

  const readCookie = (name: string): string | undefined => cookies[name];

  const sendResponse = (output: unknown) => {
    response.writeHead(200, { 'Content-Type': 'application/json' });
    response.end(JSON.stringify(output));
  };

  const sendError = (error: Error, errorCode: number) => {
    response.writeHead(errorCode, { 'Content-Type': 'application/json' });
    response.end(JSON.stringify({ error: error.message }));
  };

  const methodDetails = () => getMethodDetails(
    request.method || 'GET',
    { method: searchParams.get('method'), input: searchParams.get('input') },
    postBody,
    sendError
  );

  return {
    setCookie, readCookie, sendResponse, sendError, methodDetails
  };
};

export default nodejs;
