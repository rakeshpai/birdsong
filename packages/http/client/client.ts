/* global globalThis */
import type { RPCSerializableValue } from '../../../common/types';
import { encode, decode } from '../shared/type-handlers';

const createGetUrl = (url: string, method: string, input: RPCSerializableValue) => {
  const u = new URL(url);
  u.searchParams.set('method', method);
  u.searchParams.set('input', encode(input));
  return u.toString();
};

const canMakeGetRequest = (url: string, method: string, input: RPCSerializableValue) => (
  (method.startsWith('get')
    || method.startsWith('list'))
  && createGetUrl(url, method, input).length < 1000
);

const globalFetch = globalThis.fetch;

type ClientOptions = {
  url: string;
  fetch?: typeof globalFetch;
};

export const createClient = <T>({ url, fetch = globalFetch }: ClientOptions) => new Proxy({}, {
  get: (target, prop) => (input: RPCSerializableValue) => {
    const methodName = prop as string;
    // eslint-disable-next-line no-console
    console.log(`Invoking method ${methodName} with input:`, input);

    const fetchOptions: Parameters<typeof fetch> = canMakeGetRequest(url, methodName, input)
      ? [createGetUrl(url, methodName, input), {
        method: 'GET',
        headers: { 'Accept': 'application/json' }
      }]
      : [url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' },
        body: encode({ method: methodName, input })
      }];

    return fetch(...fetchOptions).then(res => res.text()).then(decode);
  }
}) as T;
