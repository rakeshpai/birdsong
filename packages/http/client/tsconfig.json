{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "target": "esnext",
    "lib": [
      "esnext",
      "DOM"
    ],
    "module": "ES2020",
    "moduleResolution": "Node",
    "strict": true,
    "noImplicitOverride": true,
    // Needed for CommonJS modules: markdown-it, fs-extra
    "allowSyntheticDefaultImports": true,
    //
    "sourceMap": true,
    "composite": true,
    "declaration": true,
    "declarationMap": true, // enables importers to jump to source
  },
}