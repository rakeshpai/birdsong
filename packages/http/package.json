{
  "name": "@birdsong/http",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rakeshpai/birdsong.git"
  },
  "author": "",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/rakeshpai/birdsong/issues"
  },
  "homepage": "https://github.com/rakeshpai/birdsong#readme",
  "dependencies": {
    "cookie": "^0.5.0"
  },
  "devDependencies": {
    "@types/cookie": "^0.5.1",
    "@types/node": "^17.0.36"
  }
}