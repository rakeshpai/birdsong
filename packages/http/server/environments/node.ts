/// <reference types="node" />
import type { IncomingMessage, ServerResponse } from 'http';
import cookie from 'cookie';
import { parse as urlParse } from 'url';
import type { EnvironmentHelpers } from './types';
import { getMethodDetails } from './helpers';
import type { RPCError } from '../../shared/errors';

const nodejs = (request: IncomingMessage, response: ServerResponse): EnvironmentHelpers => {
  const cookies = cookie.parse(request.headers.cookie || '');
  const searchParams = urlParse(request.url || '', true).query;

  const postBody = (() => {
    const result: Promise<string> = new Promise((resolve, reject) => {
      let body = '';
      request.on('data', chunk => { body += chunk; });
      request.on('end', () => resolve(body));
      request.on('error', reject);
    });

    return () => result;
  })();

  const setCookie = (name: string, value: string) => {
    response.setHeader('Set-Cookie', cookie.serialize(name, value));
  };

  const readCookie = (name: string): string | undefined => cookies[name];

  const sendResponse = (output: unknown) => {
    response.writeHead(200, { 'Content-Type': 'application/json' });
    response.end(output);
  };

  const sendError = (error: RPCError<any>) => {
    response.writeHead(error.statusCode || 500, { 'Content-Type': 'application/json' });
    response.end({
      error: {
        message: error.message,
        type: error.type || 'Unknown'
      }
    });
  };

  const methodDetails = () => getMethodDetails(
    request.method || 'GET',
    {
      name: searchParams.method as string | null,
      input: searchParams.input as string | null
    },
    postBody
  );

  return {
    setCookie, readCookie, sendResponse, sendError, methodDetails
  };
};

export default nodejs;
