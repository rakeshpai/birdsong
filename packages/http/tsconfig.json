{
  "extends": "../../tsconfig.json",
  "include": [
    "../../common",
    "./shared",
    "./server",
    "./client"
  ]
}