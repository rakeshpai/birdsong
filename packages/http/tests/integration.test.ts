/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable @typescript-eslint/no-redeclare */
import {
  it, beforeAll, afterAll, expect
} from 'vitest';
import { expectType } from 'ts-expect';
import http from 'http';
import fetch from 'node-fetch';
import httpServer from '../server/server';
import node from '../server/environments/node';
import { createClient } from '../client/client';

const { server: rpcServer, clientStub } = httpServer({
  environment: node,
  service: method => ({
    getUser: method(
      value => value as { id: number },
      ({ id }) => ({
        id,
        name: 'John Doe',
        dateOfBirth: new Date('1980-01-01'),
        personal: {
          interests: new Set(['Coding', 'Reading', 'Sleeping']),
          social: new Map([
            ['twitter', '@john_doe'],
            ['github', 'john_doe']
          ])
        }
      })
    ),
    getUserTypes: method(
      value => value as number,
      () => ['home', 'work', 'other']
    ),
    saveUser: method(
      value => value as { name: string },
      async ({ name }) => `'${name}' saved`
    )
  })
});

let server;

beforeAll(() => {
  server = http.createServer((req, res) => {
    if (req.url.startsWith('/api')) return rpcServer(req, res);
  }).listen(4949);
});

afterAll(() => {
  server.close();
});

type FetchType = Parameters<typeof createClient>[0]['fetch'];

it('should make an rpc call', async () => {
  const client = createClient<typeof clientStub>({
    url: 'http://localhost:4949/api',
    fetch: fetch as unknown as FetchType
  });

  const result = await client.getUser({ id: 1 });
  expectType<{
    id: number;
    name: string;
    personal: {
      interests: Set<string>;
      social: Map<string, string>;
    };
    dateOfBirth: Date;
  }>(result);
  expect(result).toMatchSnapshot();
});

it('should make another rpc call', async () => {
  const client = createClient<typeof clientStub>({
    url: 'http://localhost:4949/api',
    fetch: fetch as unknown as FetchType
  });

  const result = await client.getUserTypes(5);
  expectType<string[]>(result);
  expect(result).toMatchSnapshot();
});

it('should make yet another rpc call', async () => {
  const client = createClient<typeof clientStub>({
    url: 'http://localhost:4949/api',
    fetch: fetch as unknown as FetchType
  });

  const result = await client.saveUser({ name: 'John Doe' });
  expectType<string>(result);
  expect(result).toMatchSnapshot();
});
