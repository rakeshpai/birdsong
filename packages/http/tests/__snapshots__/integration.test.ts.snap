// Vitest Snapshot v1

exports[`should make an rpc call 1`] = `
{
  "dateOfBirth": 1980-01-01T00:00:00.000Z,
  "id": 1,
  "name": "John Doe",
  "personal": {
    "interests": Set {
      "Coding",
      "Reading",
      "Sleeping",
    },
    "social": Map {
      "twitter" => "@john_doe",
      "github" => "john_doe",
    },
  },
}
`;

exports[`should make another rpc call 1`] = `
[
  "home",
  "work",
  "other",
]
`;

exports[`should make yet another rpc call 1`] = `"'John Doe' saved"`;
